src/dev/HallSensor.cpp

```diff
@@ -0,0 +1,105 @@
+#include <EVT/io/GPIO.hpp>
+#include <EVT/utils/log.hpp>
+#include <HALf3/stm32f3xx_hal.h>
+#include <dev/HallSensor.hpp>
+
+namespace IO = EVT::core::IO;
+
+namespace WSS::DEV {
+
+HallSensor::HallSensor(IO::GPIO& gpio, uint32_t wheelRadius, uint32_t numberOfMagnets) : gpio((IO::GPIO&) gpio), wheelRadius(wheelRadius), numberOfMagnets(numberOfMagnets) {
+    this->prevTime = 0;
+    this->lastInterval = 0;
+    this->state = WheelSpeedState::STOP;
+    this->magnetInLastRead = false;
+}
+
+void HallSensor::update() {
+    uint32_t now = HAL_GetTick();
+    uint32_t elapsedTime = now - prevTime;
+
+    // The magnet is detected if the previous gpio pin reading is the same as the current gpio
+    // pin reading, and if the pin is LOW
+    const bool magnetReadLow = gpio.readPin() == MAGNET_DETECTED_STATE;
+    bool magnetDetected = false;
+    if (magnetReadLow && magnetInLastRead != magnetReadLow) {
+        magnetDetected = true;
+    }
```

Comment:
This is entirely stylistic, but I would recommend this
```suggestion
    const bool magnetDetected = magnetReadLow && !magnetInLastRead;
```